### PR TITLE
Implement api keys scope logic

### LINKED ISSUE #779: Implement api keys scope logic

### PR DESCRIPTION
**Pull request by ec**
closes #779

### CODE REVIEW

**Comment by ec** [comment]
@mod here is implementation of api keys scope logic:

### Rules

Rules fetches from `/config/authz_rules.yml` and typically should look:
```ruby
api key rules:
  scope_name: 
    - api/v2/allowed-path-for-this-scope
    - api/v2/allowed-path-for-this-scope
    ...
  scope_name:
    - api/v2/allowed-path-for-this-scope
    ...
  scope_name:
    - api/v2/allowed-path-for-this-scope
    ...
```

Although i've put the default scopes `account, trade, withdraw` inside, it can be changed by mounting custom yml in image

P.S. remember, that no need to specify the full path for every API endpoint, enough to specify module name or namespace ( as check is `starts_with?`)

### Read and Write access
Now we use `serialize :scope, JSON` in model (previously it was array)

So, now each scope can have `read` or `write` value, 
(ex. `scope: {"withdraw"=>"read"`)

Depending on the value, AuthZ will allow all the methods (POST, PUT, DELETE etc) for scopes with `write` access, and only GET requests for scope with `read` access

P.S. multiple scopes allowed for one api key.
(ex. `scope: {"withdraw"=>"read", "trade"=>"write"}`)

### Errors

As it shown in the spec in the patch, if api key tries to access the path, that is not allowed for current scope AuthZ will return error:
status: `401`
body: `{ errors: [authz.api_key_not_enough_rights] }`

The same error will be returned if scope and path fits each other, but api key trying to do a POST with READ scope access

File: app/api/v2/resource/api_keys.rb [hunk 1]
```diff
@@ -33,7 +33,7 @@ class APIKeys < Grape::API
           optional :scope,
```
Comment (ysv) [inline app/api/v2/resource/api_keys.rb]: @ec you can specify this parameter type as JSON
https://github.com/ruby-grape/grape#first-class-json-types 
This should give an ability to validate scopes on fly

File: config/authz_rules.yml [hunk 2]
```diff
@@ -2,10 +2,24 @@
 # pass for whitelisted (public) routes
 # block for blacklisted routes
 #
-rules:
+general rules:
```
Comment (ysv) [inline config/authz_rules.yml]: I would prefer underscore version `general_rules`

**Comment by ec** [comment]
Frozen

**Pull request closed without merging**